ncm-opennebula/src/main/perl/OpenNebula/Host.pm

```diff
@@ -0,0 +1,169 @@
+#${PMpre} NCM::Component::OpenNebula::Host${PMpost}
+
+use NCM::Component::OpenNebula::Commands;
+use NCM::Component::OpenNebula::Ceph;
+
+=head1 NAME
+
+NCM::Component::OpenNebula::Host adds C<KVM> hosts support to
+L<NCM::Component::OpenNebula>.
+
+=head2 Public methods
+
+=over
+
+=item manage_hosts
+
+Adds or removes C<Xen> or C<KVM> hosts.
+
+=cut
+
+sub manage_hosts
+{
+    my ($self, $one, $type, $resources, %protected) = @_;
+    my ($new, $vnm_mad);
+    my $hosts = $resources->{hosts};
+    my @existhost = $one->get_hosts();
+    my %newhosts = map { $_ => 1 } @$hosts;
+    my (@rmhosts, @failedhost);
+
+    if (exists($resources->{host_ovs}) and $resources->{host_ovs}) {
+        if ($type eq "kvm") {
+            $vnm_mad = "ovswitch";
+        } elsif ($type eq "xen") {
+            $vnm_mad = "ovswitch_brcompat";
+        }
+    } else {
+        $vnm_mad = "dummy";
+    }
+
+    foreach my $t (@existhost) {
+        # Remove the host only if there are no VMs running on it
+        if (exists($protected{$t->name})) {
+            $self->info("This resource $type is protected and can not be removed: ", $t->name);
+        } elsif (exists($newhosts{$t->name})) {
+            $self->debug(1, "We can't remove this $type host. Is required by Quattor: ", $t->name);
+        } elsif ($t->used()) {
+            $self->debug(1, "We can't remove this $type host. There are still running VMs: ", $t->name);
+        } else {
+            push(@rmhosts, $t->name);
+            $t->delete();
+        }
+    }
+
+    if (@rmhosts) {
+        $self->info("Removed $type hosts: ", join(',', @rmhosts));
+    }
+
+    foreach my $host (@$hosts) {
+        my %host_options = (
+            'name'    => $host, 
+            'im_mad'  => $type, 
+            'vmm_mad' => $type, 
+            'vnm_mad' => $vnm_mad
+        );
+        # to keep the record of our cloud infrastructure
+        # we include the host in ONE db even if it fails
+        my @hostinstances = $one->get_hosts(qr{^$host$});
+        if (scalar(@hostinstances) > 1) {
+            $self->error("Found more than one host $host. Only the first host will be modified.");
+        }
+        my $hostinstance = $hostinstances[0];
+        if (exists($protected{$host})) {
+            $self->info("This resource $type is protected and can not be created/updated: $host");
+        } elsif ($self->can_connect_to_host($host)) {
+            my $output = $self->enable_node($one, $type, $host, $resources);
```

Comment:
if you don't use `$output`, do `if ($self->enable_node($one, $type, $host, $resources)) {`